### PR TITLE
Add sagas for OnRamp to AUDIO

### PR DESCRIPTION
**Pull request by rickyrombo**
### Description

Supercedes:
- #1651 

Main updates:
- Quotes and exchanges aren't directly done via actions but instead are replaced by specific routines. Instead separate `quote` actions for USDC and SOL, there is one action that provides `purchaseInfo`, strictly information regarding the purchase of SOL for a SOL => AUDIO transaction with an estimate for how much that costs in USD (using a quote from that amount of SOL to USD.
  - This allows for us to build in padding into the quote for:
    - **Transaction Fees**: we call the `computeRoutes` on Jupiter to get the swap transactions, then compile them and get associated fees via Solana's API (this is generally a constant per transaction of 0.000005 SOL, although the Jupiter transactions can vary from 1-3 depending on how large the swap transaction gets)
    - **Associated Token Account (ATA) rent exemption fees**: Once we have the route from Jupiter, we also have the intermediate tokens needed for the swap. We check for an existing ATA on the user's wallet for each token, and if not present, we add the rent exemption minimum (calculated by calling the Solana method with 165 bytes) to the quote to cover that fee.
    - **Slippage**: To help ensure that the user does not end up with less $AUDIO than the whole number they specified, we preemptively pad the quote such that if the slippage were to bottom out (ie if slippage tolerance is set to 3% and the price slips 3%), the user would still get the expected amount of $AUDIO.
      - Slight dragon here: The quote used to generate the amount of SOL to purchase is different than the quote used to exchange that SOL for AUDIO. The price could drop more than the slippage tolerance due to this and still successfully exchange (since the first quote is separate from the second, and the second quote might have a lower AUDIO amount than the first), and the user might end up with less AUDIO. In practice, this is unlikely with a high slippage tolerance + this adjustment. A good side-effect from this is that users won't see any swap failures due to slippage, which is far worse than receiving less AUDIO.
- Likewise, instead of an action for directly executing an exchange, the exchange is wrapped up in a saga that gets triggered on an OnRamp completion.
- I added polling on wallet balances and extracted some of the pieces to the `BuyAudio.ts` audius-backend file
- There's now a much more easily followed `BuyAudioStage` enum that indicates where in the flow we are (eg. are we waiting on the user to purchase SOL? For coinbase pay's funds to show up in the wallet? For the swap to complete?)

### Dragons

*Is there anything the reviewer should be on the lookout for? Are there any dangerous changes?*

- I used Solana's API to get the real fees and ATA rent exemption minimums. This means I can do so _exactly_ requiring no buffer amount, but I still think I might add a slight buffer amount just to be _sure_ we don't fail due to some arithmetic rounding error or something.
- Speaking of arithmetic, there's a lot of different conversions happening - there's `JSBI` which is used by Jupiter, there's `BN` which is what we use, there's `u64` which is what some Solana stuff uses now (a wrapper around BN that annoyingly is lowercased so doing `new u64(..)` triggers our linter), and there's `number` which is also used by Solana. I've mixed returning SOL and lamports in this PR to some degree... could use another pass on cleaning that up probably. Ideally I'd do everything in some sort of big int type to avoid any potential float precision stuff.
- Not super dragony, but mildly infuriating: We have all sorts of different conversions in the wallet utils for different things, but they all have wildly different naming conventions and uses. These conversions could use some love...

### How Has This Been Tested?

*Please describe the tests that you ran to verify your changes. Provide repro instructions & any configuration.*

Tested E2E using a prototype BuyAudio modal as well as just by triggering sagas manually and using a phantom wallet. I also closed all my ATAs and ran it again to see the ATA fee estimate get higher and lower depending on if it was my first or second time onramping. So far have not seen any "Insufficient funds" errors.

### How will this change be monitored?

*For features that are critical or could fail silently please describe the monitoring/alerting being added.*

- [ ] TODO: Add amplitude monitoring. 
  - https://github.com/AudiusProject/audius-client/pull/1673

### Feature Flags ###

*Are all new features properly feature flagged? Describe added feature flags.*

- [ ] TODO: Add optimizely flags. (Although in the current state it's dead code)

### CODE REVIEW

File: packages/web/src/services/audius-backend/BuyAudio.ts [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+}) => {
+  await waitForLibsInit()
+  const tokenAccountInfo: AccountInfo | null =
+    await libs().solanaWeb3Manager.getAssociatedTokenAccountInfo(
+      tokenAccount.toString()
+    )
+  return tokenAccountInfo
+}
+
+export const pollForAudioBalanceChange = async ({
```
Comment (piazzatron) [inline packages/web/src/services/audius-backend/BuyAudio.ts:109]: Maybe include a comment that this throws (do we want it to throw?)

**Review by piazzatron — APPROVED** [review]
Epic PR! Really just a few minor things - let's get this merged and test the heck out of it 🚀

File: packages/web/src/common/store/ui/modals/slice.ts [hunk 2]
```diff
@@ -58,7 +59,8 @@ const initialState: ModalsState = {
   Overflow: false,
   AddToPlaylist: false,
   DeletePlaylistConfirmation: false,
-  FeatureFlagOverride: false
+  FeatureFlagOverride: false,
+  BuyAudio: true
```
[the remark this answers is not in the document]
Comment (rickyrombo) [reply packages/web/src/common/store/ui/modals/slice.ts:63]: testing purposes 😅  (shouldn't even be in here oops)

File: packages/web/src/services/audius-backend/BuyAudio.ts [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+  await waitForLibsInit()
+  return libs().solanaWeb3Manager.solanaWeb3.Keypair.fromSeed(
+    libs().Account.hedgehog.wallet.getPrivateKey()
+  ) as Keypair
+}
+
+export const getSolanaConnection = async () => {
+  await waitForLibsInit()
+  return libs().solanaWeb3Manager.connection as Connection
+}
```
[the remark this answers is not in the document]
Comment (rickyrombo) [reply packages/web/src/services/audius-backend/BuyAudio.ts:43]: yeah where should those live?

[the remark this answers is not in the document]
Comment (piazzatron) [reply packages/web/src/services/audius-backend/BuyAudio.ts:43]: Maybe `/services/audius-backend/Solana.ts`?

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/mjp-cb-onramp-sagas-v2

**Pull request merged**